lib/engine-addon.js

```diff
@@ -358,12 +356,33 @@ module.exports = {
       this.initializeAddons();
 
       var invokeArguments = args || [];
-      var addons = this.nonDuplicatedAddons();
+      var hostAddons = this.ancestorHostAddons();
 
-      return addons.map(function(addon) {
-        if (addon[methodName]) {
-          return addon[methodName].apply(addon, invokeArguments);
+      return this.addons.map(function(addon) {
+        if (!addon[methodName]) {
+          // no method to call
+          return;
+        }
+        var addonId = generateAddonId(addon);
+        var hostAddon = hostAddons[addonId];
+        if (hostAddon) {
+          switch(methodName) {
+            case 'treeFor':
+              var treeName = invokeArguments[0];
+              var hostAddonCacheKey = hostAddon.cacheKeyForTree(treeName);
+              var addonCacheKey = addon.cacheKeyForTree(treeName);
+              if (addonCacheKey != null && addonCacheKey === hostAddonCacheKey) {
+                // the addon specifies cache key and it is the same as host instance of the addon, skip the tree
+                return;
+              }
+              break;
+            case 'included':
```

Comment:
Just an explicit list of the cases, but I guess for only two cases makes for rather unordinary looking switch statement.